latest fesom2 workflow

rohanali1798 opened the pull request:
Updates of the ESM workflow will be maintained here.

What changed in the Python code (we can use that in the deliverable, maybe):

- [x] Fixed imports
- [x] Added generic installation instructions (should work on any computer or container with Internet access [PyPI])
- [x] Added logging to replace `print` calls (only stdout was used)
- [x] Removed dead code
- [x] Corrected indentation of code (parts had 2, 3, or 4 spaces)
- [x] Added argparse to handle command line arguments
- [x] Reviewed docs (Markdown)
- [x] Added Python docstrings
- [x] Added simple MyPy type hints
- [x] Reviewed string replacement in template namelists (simplified, used f-strings where possible, ...)
- [x] Hooked up the argparse parameter with the code to replace text in the template
- [x] Removed code by stdlib whenever possible (e.g. suppress)
- [x] Used newer Python 3 API where possible (discarding code to support Py2, using Pathlib, ...)
- [x] Used `getopt` in Shell scripts
- [x] Moved parameters out of the Shell script, and into the existing esm_ensemble.conf configuration file
- [x] Made shell scripts more portable (we should now be able to use this with @suvarchal on Levante -- would be nice to report this in the next deliverable)
- [x] Ran shellcheck and updated documentation with instructions on how to do that
- [ ] **Unified fesom2 and awicm3** into a single, parametrized code
- [x] shellcheck shell files
- [x] Basic unit tests before the deliverable (above 80% coverage, good enough for now)
- [x] Added a GitHub Actions CICD pipeline
- [ ] Tested on Levante with @suvarchal 

See the `README.md` file included in the esm dir for more. But the `launch_fesom2.sh` file is similar to `launch_mn4.sh` but should work on other environments too. To use it, check out this branch and:

```bash
$ ./launch_fesom2.sh \
    --hpc mn4 \
    --cores 48 \
    --start_dates "1948,1958" \
    --debug
```

Or launch it directly with Python:

```bash
$ python esm_simulation.py \
    --expid 123456 \
    --model fesom2 \
    --config ~/esm_ensemble.conf \
    --start_dates="1948,1958" \
    --debug
```

Things to clarify later:

1. The existing code has the start dates in the `esm_ensemble.conf` file, parsed via Python. Each start date is used in some "..member.." variables in the Python code (which can be quickly fixed by renaming to "..start_dates.."... but why?

https://github.com/eflows4hpc/workflow-registry/blob/b46bd34b81a2ddfb4b40fa35592eb213c6463891/Pillar_II/esm/fesom2/src/esm_simulation.py#L45C12-L45C12

![image](https://github.com/eflows4hpc/workflow-registry/assets/304786/46526c3a-5a92-448e-add9-77d4731e5d44)

https://github.com/eflows4hpc/workflow-registry/blob/b46bd34b81a2ddfb4b40fa35592eb213c6463891/Pillar_II/esm/fesom2/src/esm_ensemble_init.py#L115C16-L115C59

![image](https://github.com/eflows4hpc/workflow-registry/assets/304786/405a6453-72af-41dc-bdaf-d0ba272f7cf0)

2. The existing code has the number of members as an argument from Alien4Cloud passed to the Bash shell script `launch_mn4.sh`. That **must** match with the number of dates, as that controls how many nodes are allocated for the Slurm job. I think instead we should pass just the start dates (not sure about members, see question above), calculate the number of nodes, and then pass that to Python too (i.e. have the start dates defined a single time)

https://github.com/eflows4hpc/workflow-registry/blob/654e953c650af3e6f3122df2bf86503fe12c0145/Pillar_II/esm/fesom2/src/launch_mn4.sh#L23

rohanali1798 committed:
Rohan fixes

rohanali1798 committed:
improve output path + mpirun fix

kinow committed:
Create README.md

kinow committed:
Add some Cassandra information.

rohanali1798 committed:
Update README.md

kinow committed:
Organize imports, add a requirements.txt for dependencies, basic installation instructions. The project can now be imported into an IDE with venv.

kinow committed:
Fix indentation, re-format code (4 spaces, pep8)

kinow committed:
Clean up: remove dead code, unnecessary code, commented code, etc.

kinow committed:
Added mypy and instructions

kinow committed:
Simplify code, use stdlib suppress

kinow committed:
Making the modules more similar (meld, diff) to prepare to merge both into a single module...

kinow commented:
Alright, one full day of work (100% dedicated to eFlows4HPC :tada: ! ). I did some work on cleaning up the code first (see previous commits).

I got to a point where the `diff` of FESOM2 and AWICM3 was not too scary, and most of the old (Python2 or unused) code had been deleted or refactored to use newer Python3 API.

Then I created a new module, `esm_simulation`, with the common logic and a `main` as entry point. Added `argparse` for the command-line arguments, and finally created a `fesom2` module to encapsulate the logic that was specific for FESOM2.

I think I got most of the basic logic working for FESOM2, fully parametrized, just before it actually runs the COMPSs code. I will leave this last part for tomorrow, as well as some basic unit tests (we had zero).

Here's what it looks like:

```bash
(venv) kinow@ranma:~/Development/python/workspace/workflow-registry/Pillar_II/esm/src$ python esm_simulation.py 
INFO: Storage API successfully imported.
usage: esm_simulation [-h] [-e EXPID] -m {fesom2,awicm3} [-c CONFIG] [--debug]
esm_simulation: error: the following arguments are required: -m/--model
```

So now we have the basic help menu. You can specify the expid (`launch_mn4.sh` generates the expid in Shell and passes that to Python at the moment), or leave it blank and Python will generate an expid for you.

There is a Python enum that control what models are available. There are two at the moment, `fesom2` and `awicm3`. You must specify which one you want to run.

It will load the `esm_ensemble.conf` from the model module (`fesom2` or `awicm3`), unless you specify a `--config ~/test_ensemble.conf`. This way Rohan, myself, or anyone should be able to run the ESM code after customizing their local configuration file, without interfering with the work of others using the same ESM code.

Finally, if you pass a `--debug` the code increases the verbosity.

```bash
venv) kinow@ranma:~/Development/python/workspace/workflow-registry/Pillar_II/esm/src$ python esm_simulation.py --expid 123456 --model fesom2 --config ~/esm_ensemble.conf 
INFO: Storage API successfully imported.
INFO:__main__:Running simulation for model: fesom2
INFO:__main__:Initializing experiment 123456
INFO:__main__:Using eFlows4HPC configuration: /home/kinow/esm_ensemble.conf
INFO:__main__:New directories will be created with the file mode: 0o755
INFO:fesom2:Initializing top working directory: /tmp/eflows4hpc/top_working_dir/123456
INFO:fesom2:=== Start date 2000
INFO:fesom2:Creating symlink for FESOM2 executable, from [/tmp/eflows4hpc/fesom2/bin/fesom.x] to [/tmp/eflows4hpc/top_working_dir/123456/2000/fesom.x]
INFO:fesom2:Copying Hecuba YAML datamodel files from [/tmp/eflows4hpc/fesom2/src/io_hecuba] to the top working directory (works with .yml. YAml, .yaml, etc.)
INFO:fesom2:Generating Fortran namelists...
INFO:fesom2:=== Start date 2001
INFO:fesom2:Creating symlink for FESOM2 executable, from [/tmp/eflows4hpc/fesom2/bin/fesom.x] to [/tmp/eflows4hpc/top_working_dir/123456/2001/fesom.x]
INFO:fesom2:Copying Hecuba YAML datamodel files from [/tmp/eflows4hpc/fesom2/src/io_hecuba] to the top working directory (works with .yml. YAml, .yaml, etc.)
INFO:fesom2:Generating Fortran namelists...
INFO:fesom2:=== Start date 2002
INFO:fesom2:Creating symlink for FESOM2 executable, from [/tmp/eflows4hpc/fesom2/bin/fesom.x] to [/tmp/eflows4hpc/top_working_dir/123456/2002/fesom.x]
INFO:fesom2:Copying Hecuba YAML datamodel files from [/tmp/eflows4hpc/fesom2/src/io_hecuba] to the top working directory (works with .yml. YAml, .yaml, etc.)
INFO:fesom2:Generating Fortran namelists...
```

And debugging (it prints the before / after of namelists, it's really verbose! Use with care):

```bash
(venv) kinow@ranma:~/Development/python/workspace/workflow-registry/Pillar_II/esm/src$ python esm_simulation.py --expid 123456 --model fesom2 --config ~/esm_ensemble.conf --debug
INFO: Storage API successfully imported.
INFO:__main__:Running simulation for model: fesom2
INFO:__main__:Initializing experiment 123456
INFO:__main__:Using eFlows4HPC configuration: /home/kinow/esm_ensemble.conf
INFO:__main__:New directories will be created with the file mode: 0o755
INFO:fesom2:Initializing top working directory: /tmp/eflows4hpc/top_working_dir/123456
INFO:fesom2:=== Start date 2000
INFO:fesom2:Creating symlink for FESOM2 executable, from [/tmp/eflows4hpc/fesom2/bin/fesom.x] to [/tmp/eflows4hpc/top_working_dir/123456/2000/fesom.x]
INFO:fesom2:Copying Hecuba YAML datamodel files from [/tmp/eflows4hpc/fesom2/src/io_hecuba] to the top working directory (works with .yml. YAml, .yaml, etc.)
INFO:fesom2:Generating Fortran namelists...
DEBUG:fesom2:Processing namelist /home/kinow/Development/python/workspace/workflow-registry/Pillar_II/esm/src/fesom2/namelist_templates/namelist.config.tmpl, values {'START_YEAR': '2000', 'MESH_PATH': PosixPath('/tmp/eflows4hpc/fesom2/test/meshes/pi'), 'CLIMATOLOGY_PATH': PosixPath('/tmp/eflows4hpc/fesom2/test/input/global'), 'OUTPUT_PATH': PosixPath('/tmp/eflows4hpc/output_dir/2000')}
DEBUG:fesom2:Old namelist contents:
! This is the namelist file for model general configuration

&modelname
runid='fesom'
/

&timestep
step_per_day=96 !96 !96 !72 !72 !45 !72 !96
run_length= 1 !62 !62 !62 !28
run_length_unit='d'		! y, m, d, s
...
...
...
```

Tomorrow I will finish the FESOM2 integration moving the rest of the old code. The part missing is the one that executes COMPSs. I left it out intentionally as just importing the Python modules it already tries to connect to some socket/server. But it is not too difficult. By tomorrow I expect to have FESOM2 completed, and unit tested, ready to be tested on MN4.

-Bruno

kinow committed:
Merging FESOM2 and AWICM3 common parts, using pathlib, simplifying more the code.

kinow commented on Pillar_II/esm/src/esm_simulation.py:
This will be the main file of the ESM, the entry point called by PyCOMPSs or when we use `launch_mn4.sh`. It will use command line arguments to choose which functions to load from AWICM3 or FESOM2. These functions take care to process namelists for FESOM or AWICM3, to load other files required for their execution, prepare top working dirs, output dirs, etc.

kinow committed:
Use same PyCOMPSs version as in MN4.

kinow committed:
Enable debug option, improve logging, simplify further the code, move all pycompss workflow to new code

kinow committed:
Making the launcn_mn4.sh a more portable launch.sh, load HPC settings separately, run shellcheck over scripts

kinow committed:
Making the launcn_mn4.sh a more portable launch.sh, load HPC settings separately, run shellcheck over scripts

kinow commented:
I think it's now ready to be tested on MN4 (tomorrow). I believe if there are any issues I should be able to fix those with Jorge or Suvi's help. For today I will wrap up with docs & tests, and that's that :) It fails locally to load the `module`s, but at least it's printing some messages on the terminal now (shell code has been updated to use getopt and shellcheck'ed):

```bash
(venv) kinow@ranma:~/Development/python/workspace/workflow-registry/Pillar_II/esm/src/fesom2$ ./launch.sh --hpc mn4 --cores 48 --qos debug --members 2 --cores_per_node 48
Launching fesom2 eFlows4HPC ESM experiment...🚀

Launch arguments:

MODEL           : fesom2
HPC             : mn4
CORES           : 48
CORES PER NODE  : 48
NODES           : 2
QOS             : debug
MEMBERS         : 2

Loading mn4 configurations...
/home/kinow/Development/python/workspace/workflow-registry/Pillar_II/esm/src/fesom2/env/mn4.sh: line 6: module: command not found

```

Also moved the Hecuba configuration to the top directory of FESOM2. Before it was under `fesom2/hecuba_lib`, so users had to be aware/know about the `esm_ensemble.conf` and also the `hecuba_lib/storage_props.cfg`. Now everything is in the same directory.

kinow committed:
Move Hecuba configuration to same level as other config, shellcheck launch_mn4.sh now renamed/updated as launch.sh

kinow committed:
Pass start dates via parameters (for Alien4Cloud)

kinow committed:
Update documentation

kinow committed:
Adjust scripts after testing on MN4

kinow committed:
Handle arguments with commas (pycompss bug?) and update documentation

kinow commented:
Tested on MN4, fixed a few issues until the enqueue_compss command looked right, solved directory/paths that had to be available to the compss workflow. Things look good!

There is one error in Hecuba/COMPSs right now, so I asked for help from @jorgee and @yolandab .

```bash
While processing the following module(s):
    Module fullname   Module Filename
    ---------------   ---------------
    boost/1.64.0      /apps/modules/modulefiles/libraries/boost/1.64.0
    Hecuba/2.1_intel  /apps/HECUBA/modulefiles/Hecuba/2.1_intel.lua

unload SINGULARITY/3.7.3 (PATH) 
load SINGULARITY/3.7.3 (PATH) 
Extrae: Warning! OMP_NUM_THREADS is set but OpenMP is not supported!
Extrae: Error when setting domain for eventset 1
Extrae: Warning! OMP_NUM_THREADS is set but OpenMP is not supported!
Extrae: Error when setting domain for eventset 1
Picked up JAVA_TOOL_OPTIONS: -Xss1280k
WARNING: COMPSs Properties file is null. Setting default values
[ERRMGR]  -  FATAL:   Error loading storage configuration file: /home/***/****/.COMPSs/31278454/storage/cfgfiles/storage.properties
                      Stack trace:
                      storage.StorageException: You are trying to start a run with persistent object storage but any back-end client is loaded in the classpath.
                        at storage.StorageItf.init(StorageItf.java:45)
                        at es.bsc.compss.comm.Comm.init(Comm.java:188)
                        at es.bsc.compss.api.impl.COMPSsRuntimeImpl.<clinit>(COMPSsRuntimeImpl.java:199)

[BINDING-COMMONS] ERROR: Cannot find the COMPSsRuntimeImpl class. 
Exception in thread "main" java.lang.ExceptionInInitializerError
Caused by: java.lang.NullPointerException
        at es.bsc.compss.util.ErrorManager.fatal(ErrorManager.java:143)
        at es.bsc.compss.comm.Comm.init(Comm.java:190)
        at es.bsc.compss.api.impl.COMPSsRuntimeImpl.<clinit>(COMPSsRuntimeImpl.java:199)
```

Not sure why it's now asking for `storage.properties`. The previous `launch_mn4.sh` didn't have that, and I didn't find any difference in the compss call or other env variables.

kinow commented:
I compared the command on my `.enqueue_compss` with Rohan's same command. He's using the container, which has GCC Hecuba. But his command was not loading a `/home/***/****/.COMPSs/31278454/storage/cfgfiles/storage.properties`.

kinow commented:
Looked at the COMPSs and Hecuba code and module files (`grep`'ing, `find`'ing), but couldn't find a way to fix this issue. Will have to ask Yolanda in today's meeting.

kinow committed:
Add Python pytest unit tests, and coverage configuration

kinow committed:
Add CICD for the Pillar II ESM

kinow committed:
Enable GitHub Actions for ESM Pillar II

kinow commented:
Unit tests added with `pytest` and `coverage`. MyPy and Shellcheck, as well as `pytest`, are all now executed in a GitHub CICD pipeline, which is triggered for changes in `Pillar_II/esm` :tada: :partying_face:

kinow committed:
Add information about running the tests

kinow committed:
Do not pass an env_script to compss when not running in a container

kinow committed:
Adjust the settings for the MN4 environment, with parameters.

kinow committed:
Add the information that we had in BSC B2DROP files, so it is available to any user/reader/dev.

kinow committed:
Add the different ways to run the application (with compss and hecuba, and disabling each one)

kinow committed:
Add a TODO for AWICM3

kinow committed:
Fix how files are updated in shell (>| instead of just >). Update
namelist.oce template. Updated modules using the latest values
provided by Juanjo and @yolandab, and make sure that we always
have a trailing slash at the end of paths replaced in templates.

kinow committed:
Updating code with patch by @yolandab after eFlows4HPC GA troubleshooting session

kinow committed:
Simplify launch_fesom2.sh by moving things to a functions.sh file, add bats unit tests run by GH Actions

kinow commented:
I've committed the changes that could be directly committed, reverted on MN4, and now this is what's left to run with 2 cores:

```diff
diff --git a/Pillar_II/esm/src/fesom2/__init__.py b/Pillar_II/esm/src/fesom2/__init__.py
index d7f18c1..e604890 100644
--- a/Pillar_II/esm/src/fesom2/__init__.py
+++ b/Pillar_II/esm/src/fesom2/__init__.py
@@ -255,7 +255,7 @@ def init_output_dir(
      processes="${FESOM_CORES}",
      working_dir="{{working_dir_exe}}",
      fail_by_exit_value=True,
-     processes_per_node=48  # TODO: this probably needs to be parametrized so Levante/Juwels/LUMI can also be used?
+     processes_per_node=2  # TODO: this probably needs to be parametrized so Levante/Juwels/LUMI can also be used?
      )
 @task(
     log_file={
diff --git a/Pillar_II/esm/src/launch_fesom2.sh b/Pillar_II/esm/src/launch_fesom2.sh
index 7c53805..79a2330 100755
--- a/Pillar_II/esm/src/launch_fesom2.sh
+++ b/Pillar_II/esm/src/launch_fesom2.sh
@@ -12,7 +12,7 @@ MODEL="fesom2"
 HPC=""
 DEBUG=""
 CORES=0
-CORES_PER_NODE=48
+CORES_PER_NODE=2
 START_DATES=""
 QOS=debug

```

kinow commented:
Executed the workflow on MN4 with 2 cores (diff above), and latest commit. It worked fine, the same kind of success log files we had last week, @suvarchal , @yolandab .

![image](https://github.com/eflows4hpc/workflow-registry/assets/304786/f6a29fc9-70d0-4ba7-9ee3-265e87347ae9)

I recovered the last snapshot,

```bash
$ c4s RECOVER --qos=debug -t=00:10:00
20240117D1530h09s-31456631
```

Then tried to connect to the node with `cqlsh $node`, but it failed with

```bash
OpenBLAS blas_thread_init: RLIMIT_NPROC 379103 current, 379103 max
Traceback (most recent call last):
  File "/gpfs/apps/MN4/HECUBA/2.1/cassandra-d8tree/bin/cqlsh.py", line 161, in <module>
    from cassandra.cluster import Cluster
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/gpfs/apps/MN4/HECUBA/2.1/cassandra-d8tree/bin/../lib/cassandra-driver-internal-only-3.23.0.post0-1a184b99.zip/cassandra-driver-3.23.0.post0-1a184b99/cassandra/cluster.py", line 48, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/gpfs/apps/MN4/HECUBA/2.1/cassandra-d8tree/bin/../lib/cassandra-driver-internal-only-3.23.0.post0-1a184b99.zip/cassandra-driver-3.23.0.post0-1a184b99/cassandra/connection.py", line 38, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/gpfs/apps/MN4/HECUBA/2.1/cassandra-d8tree/bin/../lib/cassandra-driver-internal-only-3.23.0.post0-1a184b99.zip/cassandra-driver-3.23.0.post0-1a184b99/cassandra/protocol.py", line 44, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/gpfs/apps/MN4/HECUBA/2.1/cassandra-d8tree/bin/../lib/cassandra-driver-internal-only-3.23.0.post0-1a184b99.zip/cassandra-driver-3.23.0.post0-1a184b99/cassandra/cython_deps.py", line 8, in <module>
  File "/apps/PYTHON/3.6.4_ML/lib/python3.6/site-packages/numpy/__init__.py", line 142, in <module>
    from . import core
  File "/apps/PYTHON/3.6.4_ML/lib/python3.6/site-packages/numpy/core/__init__.py", line 24, in <module>
    from . import multiarray
  File "/apps/PYTHON/3.6.4_ML/lib/python3.6/site-packages/numpy/core/multiarray.py", line 14, in <module>
    from . import overrides
  File "/apps/PYTHON/3.6.4_ML/lib/python3.6/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
KeyboardInterrupt
Segmentation fault
```

I can, however, see that some NetCDF files were written to the disk. The 144 cores execution failed. Will report the details over email.

Cheers,
Bruno

kinow committed:
Finish parametrization of the Python and Shell code

kinow commented:
I will edit the latest commit until CI passes (in case it fails), and also to handle the output directory (via email @yolandab & @suvarchal pointed that looked like I was using the same directory, and FESOM2 was trying to run from a restart).

This last commit is to finish the last bits of the parametrization of the Python and Shell code. I had done the scaffolding work for parametrization, updating the old code. However, there were a few places in Python and Shell where values were hardcoded (like processes per node set to 48, FESOM_EXE fetched via environment variable, etc.). I didn't know if that was correct or not, so first I needed a clean run to test it.

I can run 2 nodes (it might be working with restarts because my first run with 2 was successful), and will use this now to validate this last commit.

These changes might be the very last changes required for this merge request! :tada: :crossed_fingers: (cc @rohanali1798 )

kinow committed:
Change PyCOMPSs parameters and order of parameters (seems to matter)

kinow committed:
Rename the variable in the output function

kinow committed:
Use a function-wrapper to pass parameters to PyCOMPSs (there is a
limitation on which @mpi parameters can be parametrized or not).

kinow committed:
Use core2, add comments about limitations found, and fix the output directory

kinow commented:
The current code is able to execute a core2 mesh 144 cores experiment, for the start date 1948 on MareNostrum4. It is fully parametrized, and produces the desired command-line (that triggers PyCOMPSs).

It triggers COMPSs, which calls Slurm `srun`, and prepares the MPI hosts and job. Everything is executed correctly, and instead of NetCDF files we get output written to Hecuba. FESOM2 reports success, and there are no errors, no segmentation faults.

Tomorrow I will modify the script to print the usage of the Shell and of the Python commands (which will be used in the next deliverable). Then have a final review of the code here, submit one more run, and if everything goes well, I will ask @rohanali1798 to test with his user. This change was made so that users need to modify only the `src/fesom2/esm_ensemble.conf` file, and then either use default values or pass parameters via the command-line.

> Unified fesom2 and awicm3 into a single, parametrized code

I am not sure if this will be done here. Probably a follow-up PR, although I am not sure if this step will be needed for the last deliverable (end of project).

> Tested on Levante with @suvarchal

I will ping Suvi tomorrow about this. If what I mentioned above works tomorrow, I will proceed to clone it and prepare the environment on Levante. Then I will wait until we have Hecuba and COMPSs ready there to launch a test on that machine.

Cheers,
Bruno

kinow committed:
Add Shell usage for getopt

kinow committed:
Complete the documentation about the Cassandra snapshots

kinow commented:
Updated instructions for Cassandra snapshots. The only test on MN4 that I couldn't complete was using more than 1 start date. Everything else worked, including confirming the data had been written to snapshots. Will ask Rohan to test that now, and wait on Suvi & Levante.

rohanali1798 committed:
improve esm_ensemble.conf

kinow reviewed (approved):
@rohanali1798 and I both tested and reviews this PR. This is now ready to be merged. Thank you!!!
